### ISSUE TITLE
Parallel requests to same URL make QWebView ignore cache policy

### ISSUE THREAD

**Issue by b3nn0** [annulen/webkit#636]
Hi,
When there are multiple WebViews executing a request to the same URL at the same time,
the request will only be executed once, and the result will then be displayed in all WebViews. This happens even if the cache policy is set to no-cache.
To Reproduce:
Create an application with multiple webviews, then load this URL with all views at the same time:
http://abatzill.de/random/uniqid.php
All views will display the same unique ID. The PHP looks like this:
```
<?php
header('Cache-Control: no-cache, no-store, must-revalidate'); // HTTP 1.1.
header('Pragma: no-cache'); // HTTP 1.0.
header('Expires: 0'); // Proxies.
echo uniqid("", true);
?>
```

I was able to fix it by - instead of blindly reusing resources - blindly reloading them. This might not be perfect either, but at least it doesn't result in WebViews "stealing" sessions from other WebViews.
To fix it, I changed `CachedResourceLoader::determineRevalidationPolicy` in `Source/WebCore/loader/cache/CachedResourceLoader.cpp` from
```
    if (existingResource->isLoading())
        return Use;
```
to
```
    if (existingResource->isLoading())
        return Reload;
```

**Comment by annulen** [comment]
Thanks for thorough analysis. Please look if applying patch fa9d2fc33f41fdfaec087ee7058a10fe3247011f is enough for your use case (it sets `Reload` for main resource)

**Comment by b3nn0** [comment]
Sadly no. Having it correct for the main resource is certainly already a step in the right direction, but I think there are still cases where this would fail.
Assume a request to some captcha service implemented in a single script - http://..../gen_captcha.php. In that case, it would be an ImageResource (at least I think so, I'm not really familiar with the code). And all WebViews/IFrames would then receive the same image, even though they shouldn't.
For other resource types (Script/CSS/Font/...) the issue is probably more theoretical.

I think the best way would probably be to always reload for the main resource as it's usually not that big, and for other resources, evaluation of the cache policy should be delayed until the first request has finished loading. Sadly, I couldn't figure out how to implement that.
But reloading is - in my opinion - always better than just using it without any evaluation.

**Comment by annulen** [comment]
I see this code above:

```c++
    // Don't reuse resources with Cache-control: no-store.
    if (existingResource->response().cacheControlContainsNoStore()) {
        LOG(ResourceLoading, "CachedResourceLoader::determineRevalidationPolicy reloading due to Cache-control: no-store.");
        logMemoryCacheResourceRequest(frame(), DiagnosticLoggingKeys::inMemoryCacheKey(), DiagnosticLoggingKeys::unusedReasonNoStoreKey());
        return Reload;
    }
```

Is it executed?

**Comment by b3nn0** [comment]
Well, it is if the resource was loaded earlier and has already finished loading.
However, when multiple requests to the same resource are done in parallel, the "first" request - which already is in the MemoryCache - has not yet finished loading, so the cache control options are not known yet. Therefore, this code path will not be taken for the "second" request. It's simply too early to know the cache control headers (as indicated by `resource->isLoading() == true`)

**Comment by b3nn0** [comment]
Hmm
In fact, I just tried your patch and it seems to work fine - even with simple captcha services that always use the same URL for captchas (I used this for testing: http://www.the-art-of-web.com/captcha_old.php )
So I assume that the captcha images are, in fact, of type MainResource. At least while loading initially.
I have no idea where the resource type comes to life. It shouldn't be possible to know it at a point where we don't know the cache policy.
Ergo: Only doing Reload for main resources is enough for me, because in practice, it is exactly the same as always doing Reloads.

Would be nice to have that patch merged into the next Alpha :)

**Comment by annulen** [comment]
Could you apply this patch with debug messages, and report result for subject url?
```diff
diff --git a/Source/WebCore/loader/cache/CachedResourceLoader.cpp b/Source/WebCore/loader/cache/CachedResourceLoader.cpp
index 35127c6470f..107f47f9ca4 100644
--- a/Source/WebCore/loader/cache/CachedResourceLoader.cpp
+++ b/Source/WebCore/loader/cache/CachedResourceLoader.cpp
@@ -79,6 +79,9 @@
 #include "Performance.h"
 #endif
 
+#include <QDebug>
+#include <QUrl>
+
 #define PRELOAD_DEBUG 0
 
 namespace WebCore {
@@ -787,6 +790,8 @@ CachedResourceLoader::RevalidationPolicy CachedResourceLoader::determineRevalida
             return Use;
     }
 
+    qDebug() << "determineRevalidationPolicy" << QUrl(existingResource->response().url()) << existingResource->response().cacheControlContainsNoStore() << existingResource->response().httpHeaderFields().isEmpty();
+
     // Don't reuse resources with Cache-control: no-store.
     if (existingResource->response().cacheControlContainsNoStore()) {
         LOG(ResourceLoading, "CachedResourceLoader::determineRevalidationPolicy reloading due to Cache-control: no-store.");
diff --git a/Source/WebCore/platform/network/qt/QNetworkReplyHandler.cpp b/Source/WebCore/platform/network/qt/QNetworkReplyHandler.cpp
index cd096bfe7f0..ad88f525e82 100644
--- a/Source/WebCore/platform/network/qt/QNetworkReplyHandler.cpp
+++ b/Source/WebCore/platform/network/qt/QNetworkReplyHandler.cpp
@@ -282,6 +282,10 @@ QNetworkReplyWrapper::QNetworkReplyWrapper(QNetworkReplyHandlerCallQueue* queue,
     connect(m_reply, SIGNAL(finished()), this, SLOT(receiveMetaData()));
     connect(m_reply, SIGNAL(readyRead()), this, SLOT(receiveMetaData()));
     connect(m_reply, SIGNAL(destroyed()), this, SLOT(replyDestroyed()));
+
+    connect(m_reply, &QNetworkReply::metaDataChanged, [this] {
+        qDebug() << "QNetworkReply::metaDataChanged" << this->m_reply->url() << this->m_reply->rawHeader("Cache-Control");
+    });
 }
 
 QNetworkReplyWrapper::~QNetworkReplyWrapper()
diff --git a/Source/WebCore/platform/network/qt/ResourceResponse.h b/Source/WebCore/platform/network/qt/ResourceResponse.h
index dba5a8d7de9..2072a15ffdf 100644
--- a/Source/WebCore/platform/network/qt/ResourceResponse.h
+++ b/Source/WebCore/platform/network/qt/ResourceResponse.h
@@ -46,6 +46,7 @@ public:
 private:
     friend class ResourceResponseBase;
 
+    void platformLazyInit(InitLevel);
     String platformSuggestedFilename() const;
 
     std::unique_ptr<CrossThreadResourceResponseData> doPlatformCopyData(std::unique_ptr<CrossThreadResourceResponseData> data) const { return data; }
diff --git a/Source/WebCore/platform/network/qt/ResourceResponseQt.cpp b/Source/WebCore/platform/network/qt/ResourceResponseQt.cpp
index e9dabd444cd..5419190d815 100644
--- a/Source/WebCore/platform/network/qt/ResourceResponseQt.cpp
+++ b/Source/WebCore/platform/network/qt/ResourceResponseQt.cpp
@@ -6,9 +6,16 @@
 #include "HTTPParsers.h"
 #include "MIMETypeRegistry.h"
 #include <QMimeDatabase>
+#include <QDebug>
+#include <QUrl>
 
 namespace WebCore {
 
+void ResourceResponse::platformLazyInit(ResourceResponseBase::InitLevel level)
+{
+    qDebug() << "platformLazyInit" << QUrl(url()) << level << httpHeaderFields().isEmpty();
+}
+
 String ResourceResponse::platformSuggestedFilename() const
 {
     // FIXME: Move to base class
```

**Comment by annulen** [comment]
@b3nn0 I've merged aforementioned patch. Are you still having troubles with parallel loading?

**Comment by b3nn0** [comment]
Sorry, only got around to do further tests today.
So, it's as I've written above, with the latest 5.212 branch.
Things are certainly better (multiple web views calling http://abatzill.de/random/uniqid.php in parallel works fine), but still not perfect (multiple web views opening a HTML page that contains something like
`<img src="http://www.the-art-of-web.com/captcha_old.php">` does not work correctly. All webviews will receive the same captcha image, even though the server sends correct headers:
```
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
```

**Comment by annulen** [comment]
Did you try to apply debug patch above and look what does it output?

**Comment by b3nn0** [comment]
Just tried it. I get a segmentation fault when starting:
platformLazyInit() calls httpHeaderFields(), which curls platformLazyInit(), ....
Infinite recursion.

**Comment by annulen** [comment]
Please try m_httpHeaderFields instead of  httpHeaderFields()

**Comment by b3nn0** [comment]
Same problem with url(). Changed it to m_url.
I then opened the app with 4 WebViews and loaded the URL in parallel:
`file:///home/adrian/coding/git.batzill.com/AutoWeb/test.html`
In there, there was the aforementioned `<img src="http://www.the-art-of-web.com/captcha_old.php">`

The resulting log:
https://paste.ubuntu.com/25738267/

**Comment by annulen** [comment]
Could yu please re-run with https://0x0.st/Cf_.txt instead of previous patch?

**Comment by b3nn0** [comment]
Same test again:
https://paste.ubuntu.com/25738791/

**Comment by annulen** [comment]
OK, so 3 `determineRevalidationPolicy` for `captcha_old.php` happen before we get first `QNetworkReply::metaDataChanged` signal so there is nothing to fix on Qt side (my thought was that we could have header available in QNetworkReplty but not yet visible to WebKit).

Looks like a thing to ask WebKit developers about.
